File: test_server.py
"""Local test server: serves a .desc file, a protobuf response (Charles CT), and a JSON response."""
import http.server
import socketserver
from google.protobuf import descriptor_pb2

# build .desc for demo.Person { string name=1; int32 id=2; }
fd = descriptor_pb2.FileDescriptorProto()
fd.name = "person.proto"
fd.package = "demo"
fd.syntax = "proto3"
msg = fd.message_type.add()
msg.name = "Person"
for name, num, typ in [("name", 1, descriptor_pb2.FieldDescriptorProto.TYPE_STRING),
                       ("id", 2, descriptor_pb2.FieldDescriptorProto.TYPE_INT32)]:
    f = msg.field.add()
    f.name = name; f.number = num; f.type = typ
    f.label = descriptor_pb2.FieldDescriptorProto.LABEL_OPTIONAL
fds = descriptor_pb2.FileDescriptorSet()
fds.file.append(fd)
DESC_BYTES = fds.SerializeToString()

# build a demo.Person message bytes (manually via the same pool)
from google.protobuf import descriptor_pool, message_factory
pool = descriptor_pool.DescriptorPool()
pool.Add(fd)
Person = message_factory.GetMessageClass(pool.FindMessageTypeByName("demo.Person"))
PB_BYTES = Person(name="Alice", id=42).SerializeToString()

PORT = 8889


class H(http.server.BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/Model.desc":
            self.send_response(200)
            self.send_header("Content-Type", "application/octet-stream")
            self.send_header("Content-Length", str(len(DESC_BYTES)))
            self.end_headers()
            self.wfile.write(DESC_BYTES)
        elif self.path == "/api/person":
            ct = 'application/x-protobuf; desc="http://127.0.0.1:%d/Model.desc"; messageType="demo.Person"' % PORT
            self.send_response(200)
            self.send_header("Content-Type", ct)
            self.send_header("Content-Length", str(len(PB_BYTES)))
            self.end_headers()
            self.wfile.write(PB_BYTES)
        elif self.path == "/api/data":
            body = b'{"game":{"id":1,"name":"TapTap"}}'
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(body)))
            self.end_headers()
            self.wfile.write(body)
        else:
            self.send_response(404)
            self.end_headers()

    def log_message(self, *a):
        pass


print(f"test server on :{PORT}", flush=True)
socketserver.TCPServer(("127.0.0.1", PORT), H).serve_forever()

File: whistle-plugin/tests/test_server.ts
/**
 * test_server.ts — mock HTTP server for E2E tests.
 *
 * Replaces the Python test_server.py — provides the same endpoints:
 *   GET /Model.desc    — returns FileDescriptorSet bytes (demo.Person)
 *   GET /api/person    — returns PB response with Charles self-describing Content-Type
 *   GET /api/data      — returns JSON response
 *
 * Uses protobufjs to build FileDescriptorSet dynamically (no .proto files).
 */

import * as http from 'http';
import protobuf from 'protobufjs';
import 'protobufjs/ext/descriptor';

const PORT = 8889;

// Build a FileDescriptorSet for demo.Person { string name = 1; int32 id = 2; }
function buildPersonDesc(): Buffer {
  const root = protobuf.Root.fromJSON({
    nested: {
      demo: {
        nested: {
          Person: {
            fields: {
              name: { type: 'string', id: 1 },
              id: { type: 'int32', id: 2 },
            }
          }
        }
      }
    }
  });
  root.resolveAll();
  const fds = (root as any).toDescriptor();
  const descExt = require('protobufjs/ext/descriptor');
  return Buffer.from(descExt.FileDescriptorSet.encode(fds).finish());
}

// Create a demo.Person message
function buildPerson(name: string, id: number): Buffer {
  const root = protobuf.Root.fromJSON({
    nested: {
      demo: {
        nested: {
          Person: {
            fields: {
              name: { type: 'string', id: 1 },
              id: { type: 'int32', id: 2 },
            }
          }
        }
      }
    }
  });
  root.resolveAll();
  const MsgType = root.lookupType('demo.Person');
  const msg = MsgType.create({ name, id });
  return Buffer.from(MsgType.encode(msg).finish());
}

const descBytes = buildPersonDesc();

const server = http.createServer((req, res) => {
  if (req.url === '/Model.desc') {
    res.writeHead(200, {
      'Content-Type': 'application/octet-stream',
      'Content-Length': descBytes.length,
    });
    res.end(descBytes);
    return;
  }

  if (req.url === '/api/person') {
    const personBytes = buildPerson('Alice', 42);
    const ct = `application/x-protobuf; desc="http://127.0.0.1:${PORT}/Model.desc"; messageType="demo.Person"`;
    res.writeHead(200, {
      'Content-Type': ct,
      'Content-Length': personBytes.length,
    });
    res.end(personBytes);
    return;
  }

  if (req.url === '/api/data') {
    const json = JSON.stringify({ game: { id: 1, name: 'TapTap' } });
    res.writeHead(200, {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(json),
    });
    res.end(json);
    return;
  }

  // POST endpoint for urlencoded body (E2E: reqRead form parsing)
  if (req.method === 'POST' && req.url === '/api/form') {
    req.on('data', () => {}); // drain
    req.on('end', () => {
      res.writeHead(200, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ received: true }));
    });
    return;
  }

  res.writeHead(404);
  res.end('not found');
});

export function startServer(): Promise<void> {
  return new Promise((resolve) => {
    server.listen(PORT, '127.0.0.1', () => {
      console.log(`[test_server] listening on http://127.0.0.1:${PORT}`);
      resolve();
    });
  });
}

export function stopServer(): Promise<void> {
  return new Promise((resolve) => server.close(() => resolve()));
}

export { PORT, descBytes };

// Run directly
if (require.main === module) {
  startServer().then(() => {
    console.log('Press Ctrl+C to stop');
  });
}
